### PR TITLE
gh-110964: Remove private _PyArg functions

### PR DESCRIPTION
**Pull request by vstinner**
Move the following private functions and structures to pycore_modsupport.h internal C API:

* _PyArg_BadArgument()
* _PyArg_CheckPositional()
* _PyArg_NoKeywords()
* _PyArg_NoPositional()
* _PyArg_ParseStack()
* _PyArg_ParseStackAndKeywords()
* _PyArg_Parser structure
* _PyArg_UnpackKeywords()
* _PyArg_UnpackKeywordsWithVararg()
* _PyArg_UnpackStack()
* _Py_ANY_VARARGS()

Changes:

* Python/getargs.h now includes pycore_modsupport.h to export functions.
* clinic.py now adds pycore_modsupport.h when one of these functions is used.
* Add pycore_modsupport.h includes when a C extension uses one of these functions.
* Define Py_BUILD_CORE_MODULE in C extensions which now include directly or indirectly (via code generated by Argument Clinic) pycore_modsupport.h:

  * _csv
  * _curses_panel
  * _dbm
  * _gdbm
  * _multiprocessing.posixshmem
  * _sqlite.row
  * _statistics
  * grp
  * resource
  * syslog

* _testcapi: bad_get() no longer uses METH_FASTCALL calling convention but METH_VARARGS. Replace _PyArg_UnpackStack() with PyArg_ParseTuple().
* _testcapi: add PYTESTCAPI_NEED_INTERNAL_API macro which is defined by _testcapi sub-modules which need the internal C API (pycore_modsupport.h): exceptions.c, float.c, vectorcall.c, watchers.c.
* Remove Include/cpython/modsupport.h header file. Include/modsupport.h no longer includes the removed header file.

* Issue: gh-110964

### CODE REVIEW

**Comment by vstinner** [comment]
This change is big because it changes 122 files generated by Argument Clinic:

```
$ git show --stat|grep 'clinic/.*.c.h'|wc -l
122
```

I prefer to write a single PR to move all private _PyArg functions at once, *because* each moved function touch many files generated by AC.

**Review by AlexWaygood — COMMENTED** [review]
The AC changes look fine to me! (I'll leave the question of whether the changes are worth making to those more skilled in C.)

**Comment by vstinner** [comment]
For clinic.py, I was lazy and used the `clinic` global variable in `bad_argument()`. The correct fix is to pass a *clinic* argument to add `bad_argument()` methods and calling sites. Problem: there are *many*. Second problem: I wrote such change many times, but I lost my work since the overall change was blocked for different reasons. Also, I tried to keep this PR as small as possible.

In short, clinic.py should be reworked later, once this change lands.

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -3500,6 +3520,8 @@ def bad_argument(self, displayname: str, expected: str, *, limited_capi: bool, e
         else:
             if expected_literal:
                 expected = f'"{expected}"'
+            if clinic is not None:
+                clinic.add_include('pycore_modsupport.h', '_PyArg_BadArgument()')
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:3524]: or perhaps this?

```suggestion
            assert clinic is not None
            clinic.add_include('pycore_modsupport.h', '_PyArg_BadArgument()')
```

Comment (vstinner) [reply Tools/clinic/clinic.py:3524]: As explained in my previous comment, I plan to write a follow-up for this code. We should not use the global variable, but pass an argument which *cannot* be None.

Comment (AlexWaygood) [reply Tools/clinic/clinic.py:3524]: Yes, I agree that's the more principled action in the longer term, and I'm fine with using the easier solution for now. But this is the same number of lines, and I think makes it clearer that we never expect `clinic` to be `None` here (if it is `None`, something has gone horribly wrong :-)

Comment (vstinner) [reply Tools/clinic/clinic.py:3524]: PR to refactor this code, to avoid `if clinic is not None`: PR #110982.

**Comment by vstinner** [comment]
> For clinic.py, I was lazy and used the clinic global variable in bad_argument().

I wrote PR #110984 for that.

**Comment by vstinner** [comment]
I merged my PR. @AlexWaygood: thanks for reviewing clinic.py changes.

**Pull request merged**